QA-1526: Terraform infrastructure for supporting TestRunner Cloud Functions.

ichengchang opened the pull request:
This PR consists of the basic infrastructure required to support the `TestRunner` streaming cloud function.

The infrastructure consist of the source bucket to store `TestRunner` results and the following resources.

- Pubsub topic for source bucket events
- IAM policy resource for publishing event notifications to topic
- Cloud Function deployer SA
- Runtime SA associated with the streaming Cloud Function

ichengchang committed:
Terraform infrastructure for supporting TestRunner Cloud Functions.

yonghaoy commented on testrunner/buckets.tf:
```diff
@@ -0,0 +1,25 @@
+# "Source" bucket for storing TestRunner results: google_service_account.testrunner_service_account[0].email creates/writes, sa_filemover reads and deletes, sa_streamer reads
+resource "google_storage_bucket" "testrunner-results-bucket" {
+  name                        = "${var.google_project}-${local.service}-results"
+  provider                    = google.target
```
nit: You may not need to specify provider in each modules. Same comments for all other modules

ichengchang committed:
Add default app engine sa membership.

ichengchang committed:
Add default app engine sa membership.

ichengchang committed:
Add default app engine sa membership.

ichengchang committed:
Add default app engine sa membership.

ichengchang committed:
Update role of TestRunner SA to BigQuery User.

rmeffan commented on testrunner/sa.tf:
```diff
@@ -33,3 +37,63 @@ resource "google_project_iam_binding" "storage_admin" {
     "serviceAccount:${google_service_account.testrunner_service_account[0].email}"
```
You can't use [google_project_iam_binding](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/google_project_iam#google_project_iam_binding) here, because it is an authoritative resource declaration. When you ran `atlantis plan -p testrunner-alpha`, it changed the member list for `roles/storage.admin` to contain only your testrunner SA.

To fix this, you need to use the  [google_project_iam_member](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/google_project_iam#google_project_iam_member) Terraform resource to modify the membership list for `roles/storage.admin` to add your testrunner SA.

ichengchang replied:
I have updated to use  `google_project_iam_member`. There are other `google_project_iam_binding` resources, should I change them too?

rmeffan reviewed (changes requested):
There is a breaking change in this PR; I am going to dimiss the approved review.

The problem has to be fixed, because it will break alpha (and presumably other Terra envs) whenever this module is deployed.

See inline comment for the description of the problem, and a potential solution.

rmeffan commented:
AP DevOps tracking issue for the downtime caused by this PR and associated PR 420 in terraform-ap-deployments is DDO-1542.

ichengchang committed:
Added storage.admin role to testrunner sa.

rmeffan reviewed (approved):
Looks good!

This is a conditional approval that is dependent on the outcome of a successful atlantis plan run for Terra alpha.

ichengchang committed:
Updated google_project_iam_binding to google_project_iam_member.

ichengchang committed:
Grant streamer storage.admin access to source bucket.

ichengchang committed:
Replaced iam_binding by iam_member reference.

ichengchang committed:
Updated google_pubsub_topic_iam_binding to google_pubsub_topic_iam_member.

rmeffan committed:
[DDO-1542] Fix Storage Admin Role Membership for Terra Envs (#204)

Refactor:
- Rename TF entities for clarity and to avoid a name collision.
- Move IAM roles to variables.tf.

Rename google_service_account to testrunner_sa.

Rename var to new terrible but descriptive name.

Workaround Part I for DDO-1542:
- Add SA resources for firecloud, leonardo, and sam.
- Add role bindings for firecloud, leonardo, and sam SAs.

Add account_id for firecloud/leonardo/sam.

Add vars for firecloud/leonardo/sam SA names.

Fix bugs, and refactor for clarity and style.

Ran terraform fmt.

Ran terraform-docs.

ichengchang committed:
Updated README. Removed --no-sort flag from terraform doc cmd.
